draft-ietf-httpbis-message-signatures.md

```diff
@@ -1283,22 +1291,122 @@ The table below contains the initial contents of the HTTP Signature Specialty Co
+
+To combat this, an application should require as much of the message as possible to be signed, within the limits of the application and deployment. The verifier should only trust message components that have been signed.
+
+## Cryptography and Signature Collision {#security-collision}
+
+The HTTP Message Signatures specification does not define any of its own cryptographic primitives, and instead relies on other specifications to define such elements. If the signature algorithm or key used to process the signature input string is vulnerable to any attacks, the resulting signature will also be susceptible to these same attacks.
+
+A common attack against signature systems is to force a signature collision, where the same signature value an be applied to multiple different inputs. Since this specification relies on reconstruction of the input string based on an HTTP message, and the list of components signed is fixed in the signature, it is difficult but not impossible for an attacker to effect such a collision. An attacker would need to manipulate the HTTP message and its covered message components in order to make the collision effective.
+
+To counter this, only vetted keys and signature algorithms should be used to sign HTTP messages. The HTTP Message Signatures Algorithm Registry is one source of potential trusted algorithms.
+
+## Key Theft {#security-keys}
+
+A foundational assumption of signature-based cryptographic systems is that the signing key is not compromised by an attacker. If the keys used to sign the message are exfiltrated or stolen, the attacker will be able to generate their own signatures using those keys. As a consequence, signers have to protect any signing key material from exfiltration, capture, and use by an attacker.
+
+## Modification of Required Message Parameters {#security-modify}
+
+An attacker could effectively deny a service by modifying an otherwise benign signature parameter or signed message component. While rejecting a modified message is the desired behavior, consistently failing signatures could lead to the verifier signature checking in order to make systems work again.
+
+If such failures are common within an application, the signer and verifier should compare their generated signature input strings with each other to determine which part of the message is being modified.
+
+## Mismatch of Signature Parameters from Message {#security-mismatch}
+
+The verifier needs to make sure that the signed message components match those in the message itself. This specification encourages this by requiring the verifier to derive these values from the message, but lazy cacheing or conveyance of the signature input string to a processing system could lead to downstream verifiers accepting a message that does not match the presented signature.
+
+## Symmetric Cryptography {#security-symmetric}
```

Comment:
In systems that allow both symmetric and asymmetric signatures in the same format, there have also been attacks where the attacker replaces, say, RSA with HMAC and uses the public key to generate a "valid" signature: https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/#RSA-or-HMAC-.

https://httpwg.org/http-extensions/draft-ietf-httpbis-message-signatures.html#name-application-of-http-message mentions this kind of consideration, but I think it should also be in the security considerations.

Reply:
That's what this section was getting at, but I think it's worth calling out this specific example as well. Thanks!